[VisBuilder] Add Capability to generate dynamic vega

ananzh opened the pull request:
### Description
In this PR, we add the capability for Visbuilder to generate dynamic Vega and Vega-Lite specifications based on user settings and aggregation configurations. Some hightlighs:
    
* developed functions `buildVegaSpecViaVega` and `buildVegaSpecViaVegaLite` that can create either Vega or Vega-Lite specifications depending on the complexity of the visualization.
* added `VegaSpec` and `VegaLiteSpec` interfaces to provide better type checking
* broken down the specification building into smaller, reusable components (like buildEncoding, buildMark, buildLegend, buildTooltip) to make the code more maintainable and easier to extend.
* added flattenDataHandler to prepare and transform data for use in Vega visualizations

Note: In this current implementation, we directly utilize the existing visualization and style configurations from Visbuilder. This approach maintains consistency with the current feature set and minimizes changes in this pull request. For future enhancements, we plan to introduce dedicated Vega-specific properties within both the visualization and style slices. This will allow us to accommodate new Vega-related settings and provide more granular control over Vega visualizations without affecting the existing Visbuilder functionality.

### Issues Resolved (Partially)
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/7067

## Screenshot

https://github.com/user-attachments/assets/4b2122b6-5e05-464c-a2ae-9bf5dd97f168

## Testing the changes


## Changelog

- feat: [VisBuilder] Add Capability to generate dynamic vega


### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

ashwin-pc reviewed (commented):
Nice job with the migration. A few comments:

1. Can you add unit tests for the change?
2. The vega based visualization has scrollbars that show up. can you fix that?
![Screenshot 2024-07-20 at 6 08 05 PM](https://github.com/user-attachments/assets/d638e7c2-f903-4945-8ee7-8dd2292f33ad)
3. When i compare the time series based charts to the legacy one i see this difference:
legacy
![image](https://github.com/user-attachments/assets/3dd94884-f97e-4e93-9093-33f14937949c)
vega:
![image](https://github.com/user-attachments/assets/a95e0d11-bc1e-4fbd-80cb-97e7fd7c2204)

4. There are a lot of type errors in the code changes. 
5. Looks like you havent fixed the area chart yet. is that for this Pr or a fast follow?
![image](https://github.com/user-attachments/assets/791283d1-af8e-4f28-a91d-4fa237d909b2)

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/build_spec_vega.ts:
Will add unit tests

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/build_vega_spec.ts:
```diff
@@ -0,0 +1,39 @@
+  visConfig: VisConfig,
+  visualization: any,
+  style: Style
+): any => {
```
will fix.

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/build_vega_spec.ts:
```diff
@@ -0,0 +1,39 @@
+ * @param {any} visConfig - The visual configuration settings.
+ * @param {VisualizationState} visualization - The visualization object (not used in this function, consider removing if unnecessary).
+ * @param {StyleState} style - The style configuration for the visualization.
+ * @returns {any} The complete Vega or Vega-Lite specification.
```
will update

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/build_vega_spec.ts:
```diff
@@ -0,0 +1,39 @@
+ * @param {StyleState} style - The style configuration for the visualization.
+ * @returns {any} The complete Vega or Vega-Lite specification.
+ */
+export const buildVegaSpec = (
```
The overall structure is similar for series and slices chart types. The detailed marks are different. If we expand chart types to support more, we could make this one to handle series and slices chart types and create special handlers for others.

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/build_vega_spec.ts:
```diff
@@ -0,0 +1,39 @@
+  const transformedData = flattenDataHandler(context, dimensions, 'series');
+
+  // Determine whether to use Vega or Vega-Lite based on the presence of split dimensions
+  if (dimensions.splitRow || dimensions.splitColumn) {
```
Currently I am seeing a dynamic counting on sub chart width is not possible in vega-lite so we use `{ name: 'chartWidth', update: 'width / splitCount - 10' },` in vega. I also see some other issues but would need more researches to confirm. 
So the answer is yes. We will revise this logic to create a func like `shouldUseVega` if we can summarize the cases in the future. 
I will add a TODO here.

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/components/encoding.ts:
Will do.

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/utils/types.ts:
```diff
@@ -0,0 +1,109 @@
+  zAxisFormat?: AxisFormat;
+}
+
+// Define a VegaLiteSpec interface
```
Will change.

ananzh commented on src/plugins/vis_builder/public/visualizations/vega/utils/types.ts:
```diff
@@ -0,0 +1,109 @@
+  };
+}
+
+// Define a more general VegaSpec interface
```
Will change.

ananzh commented on src/plugins/vis_builder/public/visualizations/vislib/area/to_expression.ts:
Sure. will do.

ananzh commented on src/plugins/vis_builder/public/visualizations/vislib/line/to_expression.ts:
sure.

ananzh commented on src/plugins/vis_builder/server/plugin.ts:
```diff
@@ -31,6 +34,21 @@ export class VisBuilderPlugin implements Plugin<VisBuilderPluginSetup, VisBuilde
+        name: i18n.translate('visbuilder.advancedSettings.visbuilderEnableVegaTitle', {
+          defaultMessage: 'Enable vega transformation in visbuilder',
+        }),
+        value: false,
```
Are you just asking the default value or you want to use vega by default in vis-builder?

ananzh commented:
> Nice job with the migration. A few comments:
> 
> 1. Can you add unit tests for the change?
> 2. The vega based visualization has scrollbars that show up. can you fix that?
>    ![Screenshot 2024-07-20 at 6 08 05 PM](https://private-user-images.githubusercontent.com/20453492/350731144-d638e7c2-f903-4945-8ee7-8dd2292f33ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NjMyODgsIm5iZiI6MTcyMTY2Mjk4OCwicGF0aCI6Ii8yMDQ1MzQ5Mi8zNTA3MzExNDQtZDYzOGU3YzItZjkwMy00OTQ1LThlZTctOGRkMjI5MmYzM2FkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE1NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1OTdlNjYzZDVkYTQwYTk2ODMxOGJlNGFmNzQ0NzlmNzNmNGE2MTI4OWMzOWQ0ZDE1MmUyZDE3NDQyNzQzMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GOcGvZ5ubzqn28-SR85wrZUK7FZvg0jLHFVYIiIN68Y)
> 3. When i compare the time series based charts to the legacy one i see this difference:
>    legacy
>    ![image](https://private-user-images.githubusercontent.com/20453492/350731353-3dd94884-f97e-4e93-9093-33f14937949c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NjMyODgsIm5iZiI6MTcyMTY2Mjk4OCwicGF0aCI6Ii8yMDQ1MzQ5Mi8zNTA3MzEzNTMtM2RkOTQ4ODQtZjk3ZS00ZTkzLTkwOTMtMzNmMTQ5Mzc5NDljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE1NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmODhmMDVkNWI1ZTFhMGI1ZDAzODA5NmNlOTVlZDUzYTAwZGUyZDQ4Zjk4Y2EzZDJiNWE3NGYxMWY3ZjYwMzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bNFn00Wbrw4QJlXChi6eB2jHEqPYIycUyTUwSr5h8mk)
>    vega:
>    ![image](https://private-user-images.githubusercontent.com/20453492/350732290-a95e0d11-bc1e-4fbd-80cb-97e7fd7c2204.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NjMyODgsIm5iZiI6MTcyMTY2Mjk4OCwicGF0aCI6Ii8yMDQ1MzQ5Mi8zNTA3MzIyOTAtYTk1ZTBkMTEtYmMxZS00ZmJkLTgwY2ItOTdlN2ZkN2MyMjA0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE1NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhNzZkYzdmNzQ3NmFjNTVkZTE4ZGM0YzM5YjQ0YTE2MWEzMjRiZTM0ZmRhMTdlNGU3MDI0YTgwY2U0MGE1ZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DVfZyXudqjVl7dVaPEm2yEUfGeC6pMPVNRSueXT5xs0)
> 4. There are a lot of type errors in the code changes.
> 5. Looks like you havent fixed the area chart yet. is that for this Pr or a fast follow?
>    ![image](https://private-user-images.githubusercontent.com/20453492/350732228-791283d1-af8e-4f28-a91d-4fa237d909b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NjMyODgsIm5iZiI6MTcyMTY2Mjk4OCwicGF0aCI6Ii8yMDQ1MzQ5Mi8zNTA3MzIyMjgtNzkxMjgzZDEtYWY4ZS00ZjI4LWE5MWQtNGZhMjM3ZDkwOWIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE1NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljOTg3MzFiMjc3NGRhOTAyZGQ0Njg0ZDc0YTRlZTJiZjdkNWJiMDk0NzczMWY3Mzg4NTJmYWQ5YjhlM2NkMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.za3MWZ_44w1p9waK54aDfVET3i4MCuRGPvLWBSJPWcc)
> 6. Did we not update the dashboard embeddable to use vega? when the toggle is on, the dashboard still seems to use vislib
>    ![image](https://private-user-images.githubusercontent.com/20453492/350732407-ce396f82-77dd-4274-9914-f8c385b41e7c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2NjMyODgsIm5iZiI6MTcyMTY2Mjk4OCwicGF0aCI6Ii8yMDQ1MzQ5Mi8zNTA3MzI0MDctY2UzOTZmODItNzdkZC00Mjc0LTk5MTQtZjhjMzg1YjQxZTdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDE1NDMwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0OGIwNjdmMGZmOTc1YjM2MDcyNWZhYjJhOTdhNDBiYWRkMjYyYWU5MjI5Yzk1NDZkN2ExNThiODQ1YjQxMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Dop2VUvXOcqcnLBprxU8doIk6FoRHIyjcDyrQNDlPJE)

* Ans to Q1: yes. will commit and add tests.
* Ans to Q2: this is not easy. I tried to add autosize fit and set height to container for some quick fixes but seems not working. I will create a sub-issue for this.
* Ans to Q3: you mean the x-axis right and the density? This is the vega's default behavior. It only show x that has y values. We didn't add any config for this part. 
* Ans to Q4: fixed in the next commit.
* Ans to Q5: fixed in the next commit.
* Ans to Q6: I haven't and will do it.

LDrago27 commented on src/plugins/vis_builder/public/visualizations/vega/components/encoding.ts:
```diff
@@ -0,0 +1,148 @@
+ * @param {any} formats - The formatting information for axes.
+ * @returns {VegaScale[]} The Vega scale configurations.
+ */
+export const buildVegaScales = (dimensions: any, formats: any): VegaScale[] => {
```
This function doesn't seem to be using the parameters **dimensions** and **formats** . If so we can clean them up.

LDrago27 commented on src/plugins/vis_builder/public/visualizations/vega/components/encoding.ts:
```diff
@@ -0,0 +1,148 @@
+ * @param {any} formats - The formatting information for axes.
+ * @returns {VegaScale[]} The Vega scale configurations.
+ */
+export const buildVegaScales = (dimensions: any, formats: any): VegaScale[] => {
```
The function doesn't seem to be using **formats** parameter. This parameter can be cleaned.

ananzh replied:
Will update. thx.

LDrago27 commented on src/plugins/vis_builder/public/visualizations/vega/utils/helpers.ts:
```diff
@@ -0,0 +1,138 @@
+ * @returns {string} The corresponding Vega data type
+ */
+export const mapFieldTypeToVegaType = (fieldType) => {
+  const typeMap = {
```
nit: We can use an enum to define the typeMap and add it to the constants file

LDrago27 commented on src/plugins/vis_builder/public/visualizations/vega/vega_lite_spec_builder.ts:
```diff
@@ -0,0 +1,89 @@
+
+  // Build the base Vega-Lite specification
+  const baseSpec: VegaLiteSpec = {
+    $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
```
nit: We can store it as a constant so it would be easier to modify it in future if required.

ananzh replied:
Will polish in a follow up PR.

LDrago27 commented on src/plugins/vis_builder/public/visualizations/vega/vega_lite_spec_builder.ts:
```diff
@@ -0,0 +1,89 @@
+    $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
+    data: { values: transformedData },
+    mark: buildMarkForVegaLite(vegaType),
+    encoding: buildVegaLiteEncoding(dimensions, formats),
```
nit: We are calling `buildVegaLiteEncoding(dimensions, formats)` three times separately in the function , we can store the value in first call and use it across places.

ananzh replied:
Will polish in a follow up PR.
